### ISSUE TITLE
Animation - Direction when LoopPingPong

### ISSUE THREAD

**Issue by SuperPoneyBoy** [mrdoob/three.js#19151]
Hi I'm testing this lovely library and i've found a bug.

I have an animation in a pingpong way.
When i click on it, the mixer.timeScale is switched between 1 and -1.
In the **first** loop click, the animation jumps straight  to the end and start going in reverse mode.
The others loops are fine.

![animation bug](https://user-images.githubusercontent.com/63805762/79483429-9bcea800-8012-11ea-8fc7-358fddf980ec.png)

Checking in the source code of three.module.js (row 445228), i'have found the problem and changed the following to make it work : 

`
_updateTime: function ( deltaTime, timeDirection ) {
**************************************^^^^^ Adding timeDirection
		var time = this.time + deltaTime;
		var duration = this._clip.duration;
		var loop = this.loop;
		var loopCount = this._loopCount;

		var pingPong = ( loop === LoopPingPong );

		if ( deltaTime === 0 ) {

			if ( loopCount === - 1 ) return time;

			return ( pingPong && ( loopCount & 1 ) === 1 ) ? duration - time : time;

		}

		if ( loop === LoopOnce ) {

			if ( loopCount === - 1 ) {

				// just started

				this._loopCount = 0;
				this._setEndings( true, true, false );

			}

			handle_stop: {

				if ( time >= duration ) {

					time = duration;

				} else if ( time < 0 ) {

					time = 0;

				} else {

					this.time = time;

					break handle_stop;

				}

				if ( this.clampWhenFinished ) this.paused = true;
				else this.enabled = false;

				this.time = time;

				this._mixer.dispatchEvent( {
					type: 'finished', action: this,
					direction: deltaTime < 0 ? - 1 : 1
				} );

			}

		} else { // repetitive Repeat or PingPong

			if ( loopCount === - 1 ) {

				// just started

				if ( deltaTime * timeDirection >= 0 ) {//updating from time direction
************************************^^^^Updating from time direction
					loopCount = 0;

					this._setEndings( true, this.repetitions === 0, pingPong );

				} else {

					// when looping in reverse direction, the initial
					// transition through zero counts as a repetition,
					// so leave loopCount at -1

					this._setEndings( this.repetitions === 0, true, pingPong );

				}

			}

			if ( time >= duration || time < 0 ) {

				// wrap around

				var loopDelta = Math.floor( time / duration ); // signed
				time -= duration * loopDelta;

				loopCount += Math.abs( loopDelta );

				var pending = this.repetitions - loopCount;

				if ( pending <= 0 ) {

					// have to stop (switch state, clamp time, fire event)

					if ( this.clampWhenFinished ) this.paused = true;
					else this.enabled = false;

					time = deltaTime > 0 ? duration : 0;

					this.time = time;

					this._mixer.dispatchEvent( {
						type: 'finished', action: this,
						direction: deltaTime > 0 ? 1 : - 1
					} );

				} else {

					// keep running

					if ( pending === 1 ) {

						// entering the last round

						var atStart = deltaTime < 0;
						this._setEndings( atStart, ! atStart, pingPong );

					} else {

						this._setEndings( false, false, pingPong );

					}

					this._loopCount = loopCount;

					this.time = time;

					this._mixer.dispatchEvent( {
						type: 'loop', action: this, loopDelta: loopDelta
					} );

				}

			} else {

				this.time = time;

			}

			if ( pingPong && ( loopCount & 1 ) === 1 ) {

				// invert time for the "pong round"

				return duration - time;

			}

		}

		return time;

	},`

In consecuent there is a need to update the caller :

`
	_update: function ( time, deltaTime, timeDirection, accuIndex ) {

		// called by the mixer

		if ( ! this.enabled ) {

			// call ._updateWeight() to update ._effectiveWeight

			this._updateWeight( time );
			return;

		}

		var startTime = this._startTime;

		if ( startTime !== null ) {

			// check for scheduled start of action

			var timeRunning = ( time - startTime ) * timeDirection;
			if ( timeRunning < 0 || timeDirection === 0 ) {

				return; // yet to come / don't decide when delta = 0

			}

			// start

			this._startTime = null; // unschedule
			deltaTime = timeDirection * timeRunning;

		}

		// apply time scale and advance time

		deltaTime *= this._updateTimeScale( time );
		var clipTime = this._updateTime( deltaTime, timeDirection );
*********************************************************^^^^ UPDATE THERE
		// note: _updateTime may disable the action resulting in
		// an effective weight of 0

		var weight = this._updateWeight( time );

		if ( weight > 0 ) {

			var interpolants = this._interpolants;
			var propertyMixers = this._propertyBindings;

			for ( var j = 0, m = interpolants.length; j !== m; ++ j ) {

				interpolants[ j ].evaluate( clipTime );
				propertyMixers[ j ].accumulate( accuIndex, weight );

			}

		}

	},
`

Can any developer update the source code ?

**Comment by Mugen87** [comment]
I'm not sure I understand the issue. Using `THREE.LoopPingPong` seems to work fine:

https://jsfiddle.net/xedtoLmy/1/

**Comment by SuperPoneyBoy** [comment]
I've updated the code : [https://jsfiddle.net/sjc3ao4m/](url)
Click on the window to change the mixer.timeScale before a loop is done and after !

`
import * as THREE from "https://threejs.org/build/three.module.js";

let renderer, scene, camera, mixer, clock;

init();
animate();

function init() {

    clock = new THREE.Clock();

    renderer = new THREE.WebGLRenderer();
    renderer.setSize(window.innerWidth, window.innerHeight);
    renderer.setPixelRatio(window.devicePixelRatio);

    document.body.appendChild(renderer.domElement);

    scene = new THREE.Scene();

    camera = new THREE.PerspectiveCamera(40, window.innerWidth / window.innerHeight, 0.1, 10);
    camera.position.set(0, 0, 5);

    const geometry = new THREE.BoxBufferGeometry();
    const material = new THREE.MeshBasicMaterial();

    const mesh = new THREE.Mesh(geometry, material);
    scene.add(mesh);

    mixer = new THREE.AnimationMixer(mesh);
    document.addEventListener("click", async (e) => {
        //Until the first loop is not finished it will jump straigth to end or to the start of animation
        //Then it will just change direction.
        mixer.timeScale *= -1;
    });

    const keyframeTrack = new THREE.VectorKeyframeTrack('.position', [0, 3], [0, 0, 0, 1, 0, 0]);
    const clip = new THREE.AnimationClip('default', - 1, [keyframeTrack]);

    const clipAction = mixer.clipAction(clip);
    clipAction.setLoop(THREE.LoopPingPong);
    clipAction.play();

}

function animate() {

    requestAnimationFrame(animate);

    mixer.update(clock.getDelta());

    renderer.render(scene, camera);

}
`

**Comment by Mugen87** [comment]
> if ( deltaTime * timeDirection >= 0 )

I don't think this modification is correct. Consider this fiddle: https://jsfiddle.net/m4th2fsc/

`AnimationMixer.timeScale` is by default `-1`. If you apply your patch, it will move from left to right (although it should move right to left like in the fiddle). 

The change does also not work if you modify `AnimationAction.timeScale`.

However, I can confirm that setting the time scale to a negative value in the first run of an action is not correct. This happens only for repetitive actions (using `THREE.LoopRepeat` or `THREE.LoopPingPong`).

**Comment by SuperPoneyBoy** [comment]
I think that in a pingpong reversed, time should start at original point and do the same animation form the same starting point.
For example, in the case of timeScale =1 the pingpong does :
t0 -> t1 -> t2 -> t3 -> t2 -> t1 -> t0
In the case of  timeScale =-1, the reversed time should be the same because the point before t0 is t1 and then is t2 ...

In the case of a LoopRepeat, i'm totaly agree because :
t0->t1->t2->t3->t0->t1->t2->t3
In the case of  timeScale =-1, the point before t0 is t3 then t2 ...

According to what i said above, the proposed patch seems to work well in both cases (LoopRepeat and LoopPingPong).
By the way i dont see any problem with the THREE.LoopRepeat, it works fine when changing timeScale to -1 at the first loop.

**Comment by Mugen87** [comment]
> I think that in a pingpong reversed, time should start at original point and do the same animation form the same starting point.

Um, this really depends on the definition. Users could rely on the current behavior and your change would break it.

> The change does also not work if you modify AnimationAction.timeScale.

It seems you have not responded to this point so far.
